File: src/main/java/ticklock/service/SynchronizedTicketPurchaseService.java
package ticklock.service;

import ticklock.domain.Event;

public class SynchronizedTicketPurchaseService implements TicketPurchaseService {

    @Override
    public boolean purchase(Event event) {
        synchronized (event) {
            if (!event.hasRemainingSeats()) {
                return false;
            }

            try {
                Thread.sleep(0, 100);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }

            event.decreaseSeat();
            return true;
        }
    }
}

File: src/main/kotlin/ticklock/service/SynchronizedTicketPurchaseService.kt
package ticklock.service

import org.springframework.stereotype.Service
import ticklock.domain.Event

@Service
class SynchronizedTicketPurchaseService : TicketPurchaseService {

    override fun purchase(event: Event): Boolean {
        synchronized(event) {
            if (!event.hasRemainingSeats()) {
                return false
            }

            try {
                Thread.sleep(0, 100)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
            }

            event.decreaseSeat()
            return true
        }
    }
}

